test: create deterministic addrman in the functional tests

stratospher opened the pull request:
An address is placed in a `[bucket,position]` in the addrman table (new table or tried table) using the `addpeeraddress` RPC. This `[bucket,position]` is calculated using `nKey`(and other metrics) for the addrman which is chosen randomly during every run.

Supposing there are 2 addresses to be placed in an addrman table. During every test run, a different `[bucket,position]` would be calculated for each address.These calculated `[bucket,position]` could even be the same for the 2 addresses in some test runs and result in collisions in the addrman. We wouldn't be able to predict when the collisions are going to happen because we can't predict the `nKey` value which is chosen at random. This can cause flaky tests.

Because of these non deterministic collisions, we are limited in what we can do to test addrman functionality. Currently in our tests don't add a second address to prevent these collisions from happening - we only keep 1 address in the new table and 1 address in the tried table. See https://github.com/bitcoin/bitcoin/pull/26988#discussion_r1091145647, https://github.com/bitcoin/bitcoin/pull/23084, [#22831(comment)](https://github.com/bitcoin/bitcoin/pull/22831/files#r708302639). 

This PR lets us create a deterministic addrman with fixed `nKey` so that we can know the `[bucket,position]` collisions beforehand, safely add more addresses in an addrman table and write more extensive tests.

stratospher committed:
[init] Remove -addrmantest command line arg

-addrmantest is only used in `p2p_node_network_limited.py` test to
test if the node self-advertises a hard-coded local address
(which wouldn't be advertised in the tests because it's unroutable
without the test-only code path) to check pruning-related services
are correct in that addr.

Remove -addrmantest because the self advertisement happens because
of hard coded test path logic, and expected services are nominal
due to how easily the test-only code could diverge from mainnet
logic. It's also being used only in 1 test.

stratospher committed:
[init] Create deterministic addrman in tests using -test=addrman

Supposing there are 2 different addresses to be placed in an addrman
table. During every test run, a different [bucket,position] would be
calculated for each address. These calculated [bucket,position] could
end up being the same for the 2 different addresses in some test runs
and result in collisions in the addrman. We wouldn't be able to
predict when the collisions are going to happen because we can't
predict the nKey value which is chosen at random. This can cause
flaky tests.

Improve this by allowing deterministic addrman creation in the
functional tests. This creates an addrman with fixed `nKey` = 1 and
we can know the [bucket,position] collisions beforehand, safely add
more addresses in an addrman table and write more extensive tests.

stratospher committed:
[test] Use deterministic addrman in addrman info tests

The nodes are restarted with an empty addrman and populated
with addresses from different networks using a helper function.
We can safely add multiple addresses to addrman tables without
worrying about unpredictable collisions since bucket:position
is fixed in a deterministic addrman.

0xB10C commented on test/functional/rpc_net.py:
```diff
@@ -305,7 +305,7 @@ def test_addpeeraddress(self):
         by first testing adding a tried table entry before testing adding a new table one.
         """
```
You might also want to edit/update/remove(?) this comment. I think mentioning that we avoid collisions by adding known-working addresses to the deterministic addrman would be good.

stratospher replied:
done.

stratospher replied:
true, i've removed the comment in commit.

brunoerg commented on test/functional/feature_asmap.py:
```diff
@@ -42,8 +42,8 @@ def set_test_params(self):
-        """Add 1 tried address to the addrman, followed by 1 new address."""
-        for addr, tried in [[0, True], [1, False]]:
+        """Add 2 tried address to the addrman, followed by 2 new address."""
+        for addr, tried in [[0, True], [1, True], [2, False], [3, False]]:
```
In 964e1a8faec984df2799171bc2a9268a3765fbca:  Any specific reason to increase the number of added addresses here?

brunoerg replied:
Thanks, @0xB10C. An explanation in the commit message would be valuable.

stratospher committed:
Revert "test: avoid non-determinism in asmap-addrman test"

This reverts commit 5825b34783545f9470d5ab94b87c918980715675.
The non-determinism is avoided by using a deterministic addrman.

stratospher commented on src/addrdb.cpp:
```diff
@@ -185,7 +185,13 @@ void ReadFromStream(AddrMan& addr, DataStream& ssPeers)
     auto check_addrman = std::clamp<int32_t>(args.GetIntArg("-checkaddrman", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), 0, 1000000);
-    auto addrman{std::make_unique<AddrMan>(netgroupman, /*deterministic=*/false, /*consistency_check_ratio=*/check_addrman)};
+
+    const auto options = gArgs.GetArgs("-addrtest");
```
done.

stratospher commented:
thanks @maflcko, @0xb10c, @brunoerg. I've updated the PR to address your comments. Open to more thoughts/suggestions in https://github.com/bitcoin/bitcoin/pull/29007#discussion_r1417082950.

0xB10C commented on test/functional/rpc_net.py:
```diff
@@ -449,6 +454,22 @@ def check_getrawaddrman_entries(expected):
                         "network": "ipv4",
                         "source": "2.0.0.0",
                         "source_network": "ipv4",
+                    },
+                    {
+                        "address": "fc00:1:2:3:4:5:6:7",
+                        "port": 8333,
+                        "services": 9,
+                        "network": "cjdns",
+                        "source": "fc00:1:2:3:4:5:6:7",
+                        "source_network": "cjdns",
+                    },
+                    {
+                        "address": "c4gfnttsuwqomiygupdqqqyy5y5emnk5c73hrfvatri67prd7vyq.b32.i2p",
+                        "port": 8333,
+                        "services": 9,
+                        "network": "i2p",
+                        "source": "c4gfnttsuwqomiygupdqqqyy5y5emnk5c73hrfvatri67prd7vyq.b32.i2p",
+                        "source_network": "i2p",
```
We're doing 1. in the [feature_addrman.py](https://github.com/bitcoin/bitcoin/blob/2e8ec6b338a825a7155fff1be83993e3834ab655/test/functional/feature_addrman.py#L160-L161) functional test.

stratospher replied:
> I think it would be ideal if we could avoid the addpeeraddress test leaking into getaddrmaninfo and getrawaddrman tests.

it would be cleaner to not need to know the existing context of addrman when dealing with getaddrmaninfo/getrawaddrman tests. both the options sound good to me!

> The easiest way of producing a collision is by adding more and more addresses to the tried table (as we do in the unit test). This would leave us with quite a few additional addresses already in the tried table.

these collisions would only affect future tests added after getaddrmaninfo/getrawaddrman tests though.

0xB10C replied:
> > The easiest way of producing a collision is by adding more and more addresses to the tried table (as we do in the unit test). This would leave us with quite a few additional addresses already in the tried table.
> 
> these collisions would only affect future tests added after getaddrmaninfo/getrawaddrman tests though.

We'd need to clean up after we found a collision.

mzumsande reviewed (commented):
Concept ACK

Needs rebase due to a silent conflict with #27581 (`feature_asmap.py`).

mzumsande commented on src/init.cpp:
```diff
@@ -579,7 +579,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-limitdescendantcount=<n>", strprintf("Do not accept transactions if any ancestor would have <n> or more in-mempool descendants (default: %u)", DEFAULT_DESCENDANT_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-limitdescendantsize=<n>", strprintf("Do not accept transactions if any ancestor would have more than <n> kilobytes of in-mempool descendants (default: %u).", DEFAULT_DESCENDANT_SIZE_LIMIT_KVB), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
-    argsman.AddArg("-addrmantest", "Allows to test address relay on localhost", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-addrtest=<option>", strprintf("Test general address related functionalities. Only used in tests. When -addrtest=relay is specified, test address relay on localhost. When -addrtest=addrman is specified, use a deterministic addrman. (default: %s)", "relay"), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Are you suggesting to have to move other existing test-only options (`-acceptstalefeeestimates` or `-fastprune` come to mind) under this command?

stratospher replied:
> It would just be a one-line function to hold the code you already wrote in two places:

oh got it now.

> I wonder if this can be generalized to pass test-only options (along with a helper HasTestOption, similar to IsDeprecatedRPCEnabled?

good idea - liked how we can make sure users don't accidentally use these test-only options in their config file.
done.

stratospher committed:
[init] Add new command line arg for use only in functional tests

some of the existing command line args are to be only used in
functional tests. ex: addrmantest, fastprune etc.. make a separate
category -test=<option> for these so that code is cleaner and
user's debug-help output is straightforward.

stratospher commented on src/net.cpp:
```diff
@@ -237,7 +237,13 @@ static int GetnScore(const CService& addr)
-    if (gArgs.GetBoolArg("-addrmantest", false)) {
+
+    const auto options = gArgs.GetArgs("-addrtest");
+    bool relayForTests = std::any_of(options.begin(), options.end(), [](const auto& option) {
```
done.

amitiuttarwar reviewed (commented):
approach ACK. yay for finally having addrman determinism in the functional tests 🙌🏽
couple thoughts to consider-

#### 1. clearer subtest addrman setup in `rpc_net.py`

@0xB10C

> I think it would be ideal if we could avoid the addpeeraddress test leaking into getaddrmaninfo and getrawaddrman tests.
> 

agreed. right now, the 3 subtests (`test_addpeeraddress`, `test_getaddrmaninfo` & `test_getrawaddrman`) have a lot of implicit dependencies, making the overall test harder to understand & develop on.

here's a suggestion with the intention of making the addrman setup more clear in each subtest:

- introduce a helper, eg. `seed_addrman` which uses `addpeeraddress` to add addresses from all the different networks to addrman.
- in the beginning of `test_getaddrmaninfo` & `test_getrawaddrman`, get a clean addrman (by using either method mentioned by @0xB10C) and call `seed_addrman` to populate it in an expected way. then each subtest would check the results with the relevant RPC

#### 2. remove `addrmantest` / `test=localaddr`

I am not convinced that this test-only behavior is worth maintaining. the only thing that `p2p_node_network_limited.py` is testing is that 1. an addr message is received & 2. the expected services are advertised. # 1 is forced to true via the test-only codepath in `GetLocalAddrForPeer`, and the value of # 2 is nominal due to how easily the test-only code could diverge from the mainnet logic.

My vote is to remove this functionality & related code. Curious to hear what you think @stratospher, as well as the opinions of other reviewers.

stratospher committed:
[test] Restart a node with empty addrman

Currently in tests where we are interested in contents of addrman,
addresses which were added to the node's addrman in previous tests
leak into the current test. example: addresses added in addpeeraddress
test leak into getaddrmaninfo and getrawaddrman tests.

It is cleaner to design the tests to be modular and without such
leaks so that we don't need to deal with context from previous tests

stratospher committed:
[test] Use deterministic addrman in addpeeraddress test

this test inserts 1 address into the new table and 1 address into
the tried table so that no collisions can happen in either table
if a second address is added. this setup does not need to be
maintained anymore since we can use a deterministic addrman and
safely add many addresses in both tables without collisions. Remove
comment explaining why previous setup needed to be maintained.

stratospher commented on test/functional/rpc_net.py:
```diff
@@ -302,13 +302,11 @@ def test_addpeeraddress(self):
         """
         self.log.info("Test addpeeraddress")
-        self.restart_node(1, ["-checkaddrman=1"])
+        self.restart_node(1, ["-checkaddrman=1", "-cjdnsreachable", "-test=addrman"])
```
makes sense! added an option in `restart_node()` to restart the node with an empty addrman which could potentially be useful in other tests. addresses added in addpeeraddress test don't leak into addrman tests now.

stratospher commented on src/init.cpp:
```diff
@@ -611,6 +611,7 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-limitdescendantcount=<n>", strprintf("Do not accept transactions if any ancestor would have <n> or more in-mempool descendants (default: %u)", DEFAULT_DESCENDANT_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-limitdescendantsize=<n>", strprintf("Do not accept transactions if any ancestor would have more than <n> kilobytes of in-mempool descendants (default: %u).", DEFAULT_DESCENDANT_SIZE_LIMIT_KVB), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-addrmantest", "Allows to test address relay on localhost", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-test=<option>", "Pass a test-only option", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
right, i've included both improvements!

stratospher commented:
thanks for the thoughtful feedback @amitiuttarwar, @0xB10C!  I've updated the PR to address your comments.

Main changes were: ([git diff](https://github.com/bitcoin/bitcoin/compare/d3c24183483e43dd38209dbaf3c4ac725e13d4a3..1b49149b3ed7366141f0090692256e82032cb9af))
1. adding an option to restart the node with an empty addrman (method 1 suggested here https://github.com/bitcoin/bitcoin/pull/29007#discussion_r1418564026) to prevent addpeeraddress test leaking into getaddrmaninfo and getrawaddrman tests.
2. listing options in -help-debug for `-test=<option>`

> I am not convinced that this test-only behavior is worth maintaining. the only thing that p2p_node_network_limited.py is testing is that 1. an addr message is received & 2. the expected services are advertised. # 1 is forced to true via the test-only codepath in GetLocalAddrForPeer, and the value of # 2 is nominal due to how easily the test-only code could diverge from the mainnet logic.

> My vote is to remove this functionality & related code. Curious to hear what you think @stratospher, as well as the opinions of other reviewers.

yes, i agree. don't think it's useful to have a command line arg which is used in only 1 line in 1 test. we would want to know if the expected services are really advertised in real world logic and not in hard coded test path logic. looks like the option was introduced in [this commit](https://github.com/stratospher/bitcoin/commit/fa999affad115c09743f8b1f5812326c19753ba9).

stratospher commented:
Updated the PR to remove `-addrmantest` ([git diff](https://github.com/bitcoin/bitcoin/compare/1b49149b3ed7366141f0090692256e82032cb9af..291a1a325db74ca5cd8ae1ca1e488ec62f99bc93)) in 5d27993.

stratospher commented on test/functional/test_framework/test_framework.py:
```diff
@@ -577,10 +577,16 @@ def stop_nodes(self, wait=0):
         self.stop_node(i)
-        self.start_node(i, extra_args)
+        if clear_addrman:
+            peers_dat = os.path.join(self.nodes[i].chain_path, "peers.dat")
```
nice! done.

0xB10C commented on src/init.cpp:
```diff
@@ -1024,6 +1024,22 @@ bool AppInitParameterInteraction(const ArgsManager& args)
 
+    if (args.IsArgSet("-test")) {
+        if (chainparams.GetChainType() != ChainType::REGTEST) {
+            return InitError(Untranslated("-test=<option> should only be used in functional tests"));
```
nit: I think this makes the error slightly clearer
```suggestion
            return InitError(Untranslated("-test=<option> can only be used when on regtest"));
```

0xB10C replied:
included in https://github.com/bitcoin/bitcoin/pull/29636

0xB10C reviewed (approved):
ACK 2cc8ca19f4185490f30a49516c890b2289fbab71

Code review and played around with the new `-test` option.

mzumsande reviewed (commented):
Code Review ACK 2cc8ca19f4185490f30a49516c890b2289fbab71

achow101 commented:
ACK 2cc8ca19f4185490f30a49516c890b2289fbab71

0xB10C commented on test/functional/rpc_net.py:
```diff
@@ -42,6 +41,20 @@ def assert_net_servicesnames(servicesflag, servicenames):
+    """ Populate the addrman with addresses from different networks.
+    Here 2 ipv4, 2 ipv6, 1 cjdns, 2 onion and 1 i2p addresses are added.
+    """
+    node.addpeeraddress(address="1.2.3.4", tried=True, port=8333)
+    node.addpeeraddress(address="2.0.0.0", port=8333)
+    node.addpeeraddress(address="1233:3432:2434:2343:3234:2345:6546:4534", tried=True, port=8333)
+    node.addpeeraddress(address="2803:0:1234:abcd::1", port=45324)
+    node.addpeeraddress(address="fc00:1:2:3:4:5:6:7", port=8333)
+    node.addpeeraddress(address="pg6mmjiyjmcrsslvykfwnntlaru7p5svn6y2ymmju6nubxndf4pscryd.onion", tried=True, port=8333)
+    node.addpeeraddress(address="nrfj6inpyf73gpkyool35hcmne5zwfmse3jl3aw23vk7chdemalyaqad.onion", port=45324, tried=True)
+    node.addpeeraddress(address="c4gfnttsuwqomiygupdqqqyy5y5emnk5c73hrfvatri67prd7vyq.b32.i2p", port=8333)
```
Ideally, we should check that these addresses are successfully added to the addrman. They work with the current deterministic addrman, but this might break silently when the addrman bucket/position calculation changes and these cause a collision.

0xB10C replied:
included in https://github.com/bitcoin/bitcoin/pull/29636

0xB10C commented on src/addrdb.cpp:
```diff
@@ -185,7 +185,9 @@ void ReadFromStream(AddrMan& addr, DataStream& ssPeers)
-    auto addrman{std::make_unique<AddrMan>(netgroupman, /*deterministic=*/false, /*consistency_check_ratio=*/check_addrman)};
+    bool deterministic = HasTestOption(args, "addrman"); // use a deterministic addrman only for tests
+
+    auto addrman{std::make_unique<AddrMan>(netgroupman, /*deterministic=*/deterministic, /*consistency_check_ratio=*/check_addrman)};
```
included in https://github.com/bitcoin/bitcoin/pull/29636
